src/charm.py

```diff
@@ -66,6 +66,74 @@ def __init__(self, *args: typing.Any):
         self.unit.open_port("tcp", 8080)  # ACL API
         self.unit.open_port("tcp", 53)  # Bind DNS
         self.unit.open_port("udp", 53)  # Bind DNS
+        self.framework.observe(self.on.create_acl_action, self._on_create_acl_action)
```

Comment:
I don't know if this is a hard requirement, but iirc we are putting the actions code into an `actions` module

Reply:
After thinking about it, I decided to use a mixin to regroup functionalities around actions. I may extract more logic from the main charm file with the same pattern going forward